File: naru_agent/orchestration/fanout.py
"""AgentFanout — parallel fan-out to multiple agents with result merging."""

from __future__ import annotations

from concurrent.futures import ThreadPoolExecutor, as_completed
from typing import Callable

from naru_agent.agent import NaruResult
from naru_agent.orchestration.orchestrator import AgentChatDelegate


class AgentFanout:
    """Parallel fan-out — send the same message to multiple agents, merge results."""

    def __init__(
        self,
        agents: list[AgentChatDelegate],
        *,
        merge: Callable[[list[NaruResult]], NaruResult] | None = None,
        max_workers: int | None = None,
        name: str = "fanout",
    ) -> None:
        if not agents:
            raise ValueError("Fan-out must have at least one agent")
        self._agents = agents
        self._merge = merge or self._default_merge
        self._max_workers = max_workers or len(agents)
        self.name = name

    def chat(
        self,
        message: str,
        user_id: str | None = None,
        session_id: str | None = None,
    ) -> NaruResult:
        with ThreadPoolExecutor(max_workers=self._max_workers) as pool:
            futures = {
                pool.submit(a.chat, message, user_id, session_id): i
                for i, a in enumerate(self._agents)
            }
            results: list[tuple[int, NaruResult]] = []
            for future in as_completed(futures):
                idx = futures[future]
                results.append((idx, future.result()))

        # Preserve original order
        results.sort(key=lambda x: x[0])
        return self._merge([r for _, r in results])

    @staticmethod
    def _default_merge(results: list[NaruResult]) -> NaruResult:
        """Default merge: concatenate content with separator."""
        contents = [r.content for r in results if r.content]
        return NaruResult(content="\n\n---\n\n".join(contents))

File: js/src/orchestration/fanout.ts
import type { NaruResult, ChatOptions } from "../types.js";
import type { AgentChatDelegate } from "./orchestrator.js";

/**
 * AgentFanout — send the same message to multiple agents in parallel,
 * then merge results.
 */
export class AgentFanout implements AgentChatDelegate {
  private mergeFn: (results: NaruResult[]) => NaruResult;

  constructor(
    private agents: AgentChatDelegate[],
    options?: {
      merge?: (results: NaruResult[]) => NaruResult;
      name?: string;
    },
  ) {
    if (agents.length === 0) {
      throw new Error("Fan-out must have at least one agent");
    }
    this.mergeFn = options?.merge ?? AgentFanout.defaultMerge;
    this.name = options?.name ?? "fanout";
  }

  public readonly name: string;

  async chat(message: string, options?: ChatOptions): Promise<NaruResult> {
    const results = await Promise.all(
      this.agents.map((a) => a.chat(message, options)),
    );
    return this.mergeFn(results);
  }

  private static defaultMerge(results: NaruResult[]): NaruResult {
    return {
      ...results[0],
      content: results
        .map((r) => r.content)
        .filter(Boolean)
        .join("\n\n---\n\n"),
    };
  }
}
